backup.sh

```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+
+set -euo pipefail
+
+# determine which environment to backup
+if [[ "${1:-}" != "testing" && "${1:-}" != "production" ]]; then
+    echo "Please provide an environment as the first argument - either \"testing\" or \"production\"!"
+    exit 1
+fi
+
+env="${1:-}"
+source .env.${env}
+
+# define place and number of backups
+mediaDirectoryInContainer="/images/uploads"
+directoryOfBackupInContainer="/backup"
+directoryOfBackup=$BACKUP_PATH
+maxNumberOfBackups=7
+
+date=$(date +%Y-%m-%d-%H-%M-%S)
+
+# create a new backup
+mkdir -m a+rw "$directoryOfBackup"/"$date"
+docker exec djangoapp-testing sqlite3 /db/db.sqlite3 ".backup '$directoryOfBackupInContainer/$date/backup_db.sqlite3'"
```

Comment:
There were some discussions whether we should stop the application for the backup or not. But I think the actual questions are how can we prevent 1) that user changes get lost and 2) we do not get an inconsistent backup. Here are some ideas with advantages/disadvantages: 

1) keep the application running and users can make changes
* backing up the database and files should work and not compromise the functionality, nor corrupt files
* however it could be that some files are copied when only partially saved so that some images would not work when loading from there (this is still pretty unlikely as the files are not that large)
* we have no downtime of the app

2) we stop the application
* user currently making changes could be interrupted and changes could get lost

3) we start a maintenance mode (show some maintenance page to the end user) and then do the backup (it should not really matter whether we stop the app or not)
* we do not lose user data
* we have a consistent state from which to make the backup 
* we do have a downtime, but it does not look as unprofessional if we have a nice maintenance page

What do you think??

Reply:
Why is this necessary at all? Are you sure the sqlite backup tool does not handle concurrency such as locking the database waiting for transactions to end etc. From [this](https://stackoverflow.com/questions/25675314/how-to-backup-sqlite-database) it sounds to me like it does. There is more info on backing up sqlite [here](https://sqlite.org/backup.html) (which I did not read ;-)). If the tool handles this properly, all the precautions mentioned above are not necessary, are they?

If I am mistaken and this is not handled by the backup tool, I'd say let's be pragmatic, use option 2 and run the backup at 3am. Not likely to have anyone working on the database at that time. Once we have got hundreds of users, this can be reassessed.

Reply:
That's what I wrote: the db backup should be no problem at all when the app is running. There is just the point that uploaded image files might be missing (but I thinks it's unlikely). 
Option 2 might be worse, I don't know how gracefully docker shuts down the app and what happens with scheduled transactions.

Reply:
As long as we use docker stop and docker compose down, it waits a few seconds for the app to terminate gracefully, given that our container handles signals correctly (I think it does, since the container usually stops quickly before the timeout).

But yes ideally sqlite handles all that and we can keep the app running. To reduce the risk for the images we could still create the backup at 3 or 4 am when people won't be uploading much.

Reply:
It was decided to go with option 1 for now.